### PR TITLE
Close selector after write

### PR DESCRIPTION
**Pull request by shvo123**
following File Descriptor leak issue #3705, we added slector.close() and this leak stopped. Closing/destroying ChannelOutputStream object does not close the selector therefore it retains redundant pipes/FD that cen be seen using lsof command or ls /proc/<pid>

### CODE REVIEW

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler - Copy.txt [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+ * Tcp outbound channel adapter using a TcpConnection to
+ * send data - if the connection factory is a server
+ * factory, the TcpListener owns the connections. If it is
+ * a client factory, this object owns the connection.
+ *
+ * @author Gary Russell
+ * @author Artem Bilan
+ * @since 2.0
+ */
+public class TcpSendingMessageHandler extends AbstractMessageHandler implements
```
Comment (garyrussell) [inline spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler - Copy.txt:53]: What is this file for?

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler.java [hunk 2]
```diff
@@ -47,9 +48,7 @@
  *
  * @author Gary Russell
  * @author Artem Bilan
- *
  * @since 2.0
- *
  */
```
Comment (garyrussell) [inline spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler.java:52]: Please don't reformat Javadocs.

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler.java [hunk 3]
```diff
@@ -58,7 +57,7 @@
 	 * A default retry interval for the {@link ClientModeConnectionManager} rescheduling.
 	 */
 	public static final long DEFAULT_RETRY_INTERVAL = 60000;
-
+	private static final int MAXIMUM_CONNECTIONS = 60;
```
Comment (garyrussell) [inline spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler.java:60]: We require empty lines between fields.

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler.java [hunk 4]
```diff
@@ -84,9 +83,10 @@ protected TcpConnection obtainConnection(Message<?> message) {
 		Assert.notNull(this.clientConnectionFactory, "'clientConnectionFactory' cannot be null");
 		try {
 			connection = this.clientConnectionFactory.getConnection();
-		}
-		catch (Exception ex) {
-			logger.error(ex, "Error creating connection");
+		} catch (Exception ex) {
```
Comment (garyrussell) [inline spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler.java:86]: Don't change formatting. Run `./gradlew check` to get checkstyle errors.

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler.java [hunk 5]
```diff
@@ -95,14 +95,14 @@ protected TcpConnection obtainConnection(Message<?> message) {
 	/**
 	 * Writes the message payload to the underlying socket, using the specified
 	 * message format.
+	 *
```
Comment (garyrussell) [inline spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler.java:98]: Don't reformat javadocs.

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler.java [hunk 6]
```diff
@@ -95,14 +95,14 @@ protected TcpConnection obtainConnection(Message<?> message) {
...
 	 * @see org.springframework.messaging.MessageHandler#handleMessage(org.springframework.messaging.Message)
 	 */
 	@Override
 	public void handleMessageInternal(final Message<?> message) {
 		if (this.serverConnectionFactory != null) {
 			handleMessageAsServer(message);
-		}
-		else {
+		} else {
```
Comment (garyrussell) [inline spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler.java:105]: It's hard to see what changed with all this noise.

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler.java [hunk 7]
```diff
@@ -226,11 +218,38 @@ public void setConnectionFactory(AbstractConnectionFactory connectionFactory) {
 	@Override
 	public void addNewConnection(TcpConnection connection) {
 		this.connections.put(connection.getConnectionId(), connection);
+
+		logger.debug(System.currentTimeMillis() + "," + LocalDateTime.now() + ",add," + connection.getConnectionId() + "," + this.connections.size() + "," +
+				this.hashCode());
+		int leakageCount = 0;
+		if (connections.size() > MAXIMUM_CONNECTIONS) {
```
Comment (garyrussell) [inline spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler.java:225]: We will not be using this change; I will fix the `AbstractConnectionFactory` as discussed in https://github.com/spring-projects/spring-integration/issues/3701#issuecomment-1003135701

**Review by garyrussell — CHANGES_REQUESTED** [review]
Thanks for the contribution; the `selector.close()` fix looks correct but we can't accept all those other unnecessary changes.

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler - Copy.txt (hunk 1 above)
Comment (shvo123) [reply spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler - Copy.txt:53]: sorry it is not required at all. my mistake.

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler.java (hunk 7 above)
Comment (shvo123) [reply spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpSendingMessageHandler.java:225]: OK

**Comment by shvo123** [comment]
would create new pull request with just the selector.close()

**Pull request closed without merging**